### ISSUE TITLE
Provider with HTTPBasic Authentication - Unable to fetch token

### ISSUE THREAD

**Issue by pgiraud** [singingwolfboy/flask-dance#233]
I want my application to have the users being authenticated via a gitlab instance we host. Any request to this server requires basic http authentication.

After the user has authorized the application on gitlab, he's been redirected and the `oauth lib` tries to fetch the token. The problem is that without valid credentials the returned response is 401 and I can't get any further in the authentication process.

I've already tried to provide the session an `auth` property but it gets overwritten in `requests_oauthlib.oauth2_session.OAuth2Session.fetch_token`. https://github.com/requests/requests-oauthlib/blob/master/requests_oauthlib/oauth2_session.py#L302

If I provide my own tuple or `requests.auth.HTTPBasicAuth` as value for `auth` when calling ` fetch_token`, I then get an `InvalidClientError` which apparently means that the basic authentification is OK, but unfortunately `client_id` seem to be missing.

Am I missing something?
Thanks for any help.

**Comment by singingwolfboy** [comment]
@jgeorgeson, you're the maintainer for the GitLab configuration. Can you help out @pgiraud at all?

**Comment by pgiraud** [comment]
I don't think this is a gitlab related issue. If use the generic `OAuth2ConsumerBlueprint` I get the same problems.

**Comment by singingwolfboy** [comment]
I believe that you can only assign one authentication object to a particular request. Requests has a built-in [HTTPBasicAuth](https://github.com/kennethreitz/requests/blob/master/requests/auth.py#L79) class, and `requests-oauthlib` defines an [OAuth2](https://github.com/requests/requests-oauthlib/blob/master/requests_oauthlib/oauth2_auth.py#L7) class, which is what Flask-Dance uses. Maybe you need to define your own auth class that inherits from both classes, and use that?

Unfortunately, I haven't come across this situation before, so I don't know if this will actually work, it's just a guess. Give it a try, and let me know if it works?

**Comment by daenney** [comment]
This seems very odd though. If I'm reading this right what you're saying is that to fetch the token, after successful authorization, we still need a separate set of credentials? Normally once authorization is completed you get a token, and that's what you can then use to do any follow-up requests.

**Comment by pgiraud** [comment]
The use case is a bit odd I agree.
In my case, the Basic Auth is provided by Apache while Oauth is managed by gitlab itself.
When my application redirects me to gitlab to get authorized, I'm asked for credentials if I haven't visited gitlab before.
But the `fetch_token` request sent by flask_dance also needs credentials to access gitlab oauth services.
I know, this extra layer is annoying. I should probably ask the sys admins for some guidance or have them unsecure the oauth services.

**Comment by daenney** [comment]
Giving it a bit more thought, I don't think this can ever work. In order to pass basic authentication you have to set a header of `Authorization: Basic <base64 thingy>` but you'll also have to pass a header of `Authorization: Bearer <token>` in order to be able to get to different Gitlab endpoints.

This would mean that Flask-Dance now needs to generate a header of `Authorization: Bearer <token>, Basic <base64 thingy>` and hope that both Apache and Gitlab don't trip over having multiple values in the header. Officially the HTTP spec supports it, but in reality I doubt anyone has implemented their code in such a way that that works.

**Comment by jgeorgeson** [comment]
@pgiraud I would have to concur with @daeney on this. If your using are authenticating to your application with Basic, what is the GitLab OAUTH token being used for? (or vice versa).

**Comment by pgiraud** [comment]
This may seem weird indeed. First of all, the basic auth is an extra security layer since we have a lot of people working remotely (including me). This authentication is spread over a large number of services.
So when I want to open gitlab in my browser, I have to authenticate twice. The main http auth and then the gitlab's one.

Here's the reason why I need the OAuth token: I would like to create an application which uses the gitlab API to show issues and merge requests in a simplified and topic oriented interface. I need OAuth authentication so that I can then use the `python-gitlab` library to access to the API with the `oauth_token`. With this token the application cand send requests to the API on behalf of the user.

 Note: I've tried the `python-gitlab` CLI with a personal token (not OAuth) along with HTTP authentication (`http_username` and `http_password`) and it just works.
https://python-gitlab.readthedocs.io/en/stable/cli.html#id3
https://python-gitlab.readthedocs.io/en/stable/api/gitlab.html?highlight=http_username#gitlab.Gitlab

But looking more deeply in the `python-gitlab`'s code, it's mentioned that we can't use both http or oauth token authentication.
https://github.com/python-gitlab/python-gitlab/blob/57fa4e37aaf6ccee0d75085520f96fd15752a3df/gitlab/__init__.py#L342-L344

The conflict on headers as mentioned by @daenney makes a lot of sense.
Thanks for your help. I really need to talk with the sys admin. Closing.
